P2P: Cannot start a transaction within a transaction

##### android.database.sqlite.SQLiteException: cannot start a transaction within a transaction (code 1)

This exception is thrown at listener side.

Steps to reproduce:
1. Create 100 docs on listener side and 100 docs on client side same time.

```
03-11 14:19:35.825  29324-29717/couchbase.com.myp2p E/Router﹕ Router unable to route request to do_POST_Document_bulk_docs
    java.lang.reflect.InvocationTargetException
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.couchbase.lite.router.Router.start(Router.java:472)
            at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:128)
            at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
            at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2251)
            at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2176)
            at Acme.Serve.Serve$ServeConnection.run(Serve.java:1988)
            at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1223)
            at java.lang.Thread.run(Thread.java:818)
     Caused by: android.database.sqlite.SQLiteException: cannot start a transaction within a transaction (code 1)
            at android.database.sqlite.SQLiteConnection.nativeExecute(Native Method)
            at android.database.sqlite.SQLiteConnection.execute(SQLiteConnection.java:555)
            at android.database.sqlite.SQLiteSession.beginTransactionUnchecked(SQLiteSession.java:323)
            at android.database.sqlite.SQLiteSession.beginTransaction(SQLiteSession.java:298)
            at android.database.sqlite.SQLiteDatabase.beginTransaction(SQLiteDatabase.java:505)
            at android.database.sqlite.SQLiteDatabase.beginTransaction(SQLiteDatabase.java:416)
            at com.couchbase.lite.android.AndroidSQLiteStorageEngine.beginTransaction(AndroidSQLiteStorageEngine.java:78)
            at com.couchbase.lite.Database.beginTransaction(Database.java:1253)
            at com.couchbase.lite.router.Router.do_POST_Document_bulk_docs(Router.java:1080)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.couchbase.lite.router.Router.start(Router.java:472)
            at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:128)
            at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
            at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2251)
            at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2176)
            at Acme.Serve.Serve$ServeConnection.run(Serve.java:1988)
            at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1223)
            at java.lang.Thread.run(Thread.java:818)
    --------- beginning of crash
03-11 14:19:35.840  29324-29324/couchbase.com.myp2p E/AndroidRuntime﹕ FATAL EXCEPTION: main
    Process: couchbase.com.myp2p, PID: 29324
    java.lang.IllegalStateException: Cannot perform this operation because there is no current transaction.
            at android.database.sqlite.SQLiteSession.throwIfNoTransaction(SQLiteSession.java:915)
            at android.database.sqlite.SQLiteSession.setTransactionSuccessful(SQLiteSession.java:371)
            at android.database.sqlite.SQLiteDatabase.setTransactionSuccessful(SQLiteDatabase.java:540)
            at com.couchbase.lite.android.AndroidSQLiteStorageEngine.setTransactionSuccessful(AndroidSQLiteStorageEngine.java:88)
            at com.couchbase.lite.Database.endTransaction(Database.java:1288)
            at com.couchbase.lite.Database.putRevision(Database.java:3775)
            at com.couchbase.lite.Database.putRevision(Database.java:3548)
            at com.couchbase.lite.Document.putProperties(Document.java:417)
            at com.couchbase.lite.Document.putProperties(Document.java:252)
            at couchbase.com.myp2p.MainActivity$2.onClick(MainActivity.java:101)
            at android.view.View.performClick(View.java:4756)
            at android.view.View$PerformClick.run(View.java:19749)
            at android.os.Handler.handleCallback(Handler.java:739)
            at android.os.Handler.dispatchMessage(Handler.java:95)
            at android.os.Looper.loop(Looper.java:135)
            at android.app.ActivityThread.main(ActivityThread.java:5221)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
```

--- a/src/main/java/com/couchbase/lite/Database.java
+++ b/src/main/java/com/couchbase/lite/Database.java
@@ -56,7 +56,11 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
-import java.util.concurrent.*;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 
 /**
@@ -75,12 +79,14 @@ public final class Database {
 
     private boolean open = false;
 
-    static class TransactionLevel extends ThreadLocal<Integer>{
+    // transactionLevel is per thread
+    static class TransactionLevel extends ThreadLocal<Integer> {
         @Override
         protected Integer initialValue() {
             return 0;
         }
     }
+
     private TransactionLevel transactionLevel = new TransactionLevel();
 
 
@@ -1210,7 +1216,6 @@ public final class Database {
             database.close();
         }
         open = false;
-        //transactionLevel = 0;
         return true;
     }
 
@@ -1281,25 +1286,19 @@ public final class Database {
      */
     @InterfaceAudience.Private
     public boolean beginTransaction() {
-
         int tLevel = transactionLevel.get();
-
-        Log.e(Log.TAG_DATABASE, "[Database.beginTransaction()] Database="+this + ", SQLiteStorageEngine="+database + ", transactionLevel="+transactionLevel + ", currentThread="+Thread.currentThread());
+        Log.d(Log.TAG_DATABASE, "[Database.beginTransaction()] Database=" + this + ", SQLiteStorageEngine=" + database + ", transactionLevel=" + transactionLevel + ", currentThread=" + Thread.currentThread());
         try {
             // Outer (level 0)  transaction. Use SQLiteDatabase.beginTransaction()
-            if(tLevel == 0) {
+            if (tLevel == 0) {
                 database.beginTransaction();
             }
             // Inner (level 1 or higher) transaction. Use SQLite's SAVEPOINT
-            else{
+            else {
                 database.execSQL("SAVEPOINT cbl_" + Integer.toString(tLevel));
             }
-
             Log.i(Log.TAG, "%s Begin transaction (level %d)", Thread.currentThread().getName(), tLevel);
-
-            ++tLevel;
-            transactionLevel.set(tLevel);
-
+            transactionLevel.set(++tLevel);
         } catch (SQLException e) {
             Log.e(Database.TAG, Thread.currentThread().getName() + " Error calling beginTransaction()", e);
             return false;
@@ -1317,13 +1316,12 @@ public final class Database {
     public boolean endTransaction(boolean commit) {
         int tLevel = transactionLevel.get();
 
-        assert(tLevel > 0);
+        assert (tLevel > 0);
 
-        --tLevel;
-        transactionLevel.set(tLevel);
+        transactionLevel.set(--tLevel);
 
         // Outer (level 0) transaction. Use SQLiteDatabase.setTransactionSuccessful() and SQLiteDatabase.endTransaction()
-        if(tLevel == 0) {
+        if (tLevel == 0) {
             if (commit) {
                 Log.i(Log.TAG, "%s Committing transaction (level %d)", Thread.currentThread().getName(), tLevel);
                 database.setTransactionSuccessful();
@@ -1339,7 +1337,7 @@ public final class Database {
             }
         }
         // Inner (level 1 or higher) transaction: Use SQLite's ROLLBACK and RELEASE
-        else{
+        else {
             if (commit) {
                 Log.i(Log.TAG, "%s Committing transaction (level %d)", Thread.currentThread().getName(), tLevel);
             } else {
@@ -1351,10 +1349,9 @@ public final class Database {
                     return false;
                 }
             }
-            try{
+            try {
                 database.execSQL("RELEASE cbl_" + Integer.toString(tLevel));
-            }
-            catch (SQLException e) {
+            } catch (SQLException e) {
                 Log.e(Database.TAG, Thread.currentThread().getName() + " Error calling endTransaction()", e);
                 return false;
             }
@@ -1645,7 +1642,7 @@ public final class Database {
      */
     @InterfaceAudience.Private
     public RevisionInternal loadRevisionBody(RevisionInternal rev, EnumSet<TDContentOptions> contentOptions) throws CouchbaseLiteException {
-        if(rev.getBody() != null && contentOptions == EnumSet.noneOf(Database.TDContentOptions.class) && rev.getSequence() != 0) {
+        if(rev.getBody() != null && contentOptions == EnumSet.noneOf(TDContentOptions.class) && rev.getSequence() != 0) {
             return rev;
         }
 

--- a/src/main/java/com/couchbase/lite/Database.java
+++ b/src/main/java/com/couchbase/lite/Database.java
@@ -74,7 +74,15 @@ public final class Database {
     private SQLiteStorageEngine database;
 
     private boolean open = false;
-    private int transactionLevel = 0;
+
+    static class TransactionLevel extends ThreadLocal<Integer>{
+        @Override
+        protected Integer initialValue() {
+            return 0;
+        }
+    }
+    private TransactionLevel transactionLevel = new TransactionLevel();
+
 
     /**
      * @exclude
@@ -1202,7 +1210,7 @@ public final class Database {
             database.close();
         }
         open = false;
-        transactionLevel = 0;
+        //transactionLevel = 0;
         return true;
     }
 
@@ -1273,19 +1281,24 @@ public final class Database {
      */
     @InterfaceAudience.Private
     public boolean beginTransaction() {
+
+        int tLevel = transactionLevel.get();
+
+        Log.e(Log.TAG_DATABASE, "[Database.beginTransaction()] Database="+this + ", SQLiteStorageEngine="+database + ", transactionLevel="+transactionLevel + ", currentThread="+Thread.currentThread());
         try {
             // Outer (level 0)  transaction. Use SQLiteDatabase.beginTransaction()
-            if(transactionLevel == 0) {
+            if(tLevel == 0) {
                 database.beginTransaction();
             }
             // Inner (level 1 or higher) transaction. Use SQLite's SAVEPOINT
             else{
-                database.execSQL("SAVEPOINT cbl_" + Integer.toString(transactionLevel));
+                database.execSQL("SAVEPOINT cbl_" + Integer.toString(tLevel));
             }
 
-            Log.i(Log.TAG, "%s Begin transaction (level %d)", Thread.currentThread().getName(), transactionLevel);
+            Log.i(Log.TAG, "%s Begin transaction (level %d)", Thread.currentThread().getName(), tLevel);
 
-            ++transactionLevel;
+            ++tLevel;
+            transactionLevel.set(tLevel);
 
         } catch (SQLException e) {
             Log.e(Database.TAG, Thread.currentThread().getName() + " Error calling beginTransaction()", e);
@@ -1302,19 +1315,21 @@ public final class Database {
      */
     @InterfaceAudience.Private
     public boolean endTransaction(boolean commit) {
+        int tLevel = transactionLevel.get();
 
-        assert(transactionLevel > 0);
+        assert(tLevel > 0);
 
-        --transactionLevel;
+        --tLevel;
+        transactionLevel.set(tLevel);
 
         // Outer (level 0) transaction. Use SQLiteDatabase.setTransactionSuccessful() and SQLiteDatabase.endTransaction()
-        if(transactionLevel == 0) {
+        if(tLevel == 0) {
             if (commit) {
-                Log.i(Log.TAG, "%s Committing transaction (level %d)", Thread.currentThread().getName(), transactionLevel);
+                Log.i(Log.TAG, "%s Committing transaction (level %d)", Thread.currentThread().getName(), tLevel);
                 database.setTransactionSuccessful();
                 database.endTransaction();
             } else {
-                Log.i(Log.TAG, "%s CANCEL transaction (level %d)", Thread.currentThread().getName(), transactionLevel);
+                Log.i(Log.TAG, "%s CANCEL transaction (level %d)", Thread.currentThread().getName(), tLevel);
                 try {
                     database.endTransaction();
                 } catch (SQLException e) {
@@ -1326,18 +1341,18 @@ public final class Database {
         // Inner (level 1 or higher) transaction: Use SQLite's ROLLBACK and RELEASE
         else{
             if (commit) {
-                Log.i(Log.TAG, "%s Committing transaction (level %d)", Thread.currentThread().getName(), transactionLevel);
+                Log.i(Log.TAG, "%s Committing transaction (level %d)", Thread.currentThread().getName(), tLevel);
             } else {
-                Log.i(Log.TAG, "%s CANCEL transaction (level %d)", Thread.currentThread().getName(), transactionLevel);
+                Log.i(Log.TAG, "%s CANCEL transaction (level %d)", Thread.currentThread().getName(), tLevel);
                 try {
-                    database.execSQL(";ROLLBACK TO cbl_" + Integer.toString(transactionLevel));
+                    database.execSQL(";ROLLBACK TO cbl_" + Integer.toString(tLevel));
                 } catch (SQLException e) {
                     Log.e(Database.TAG, Thread.currentThread().getName() + " Error calling endTransaction()", e);
                     return false;
                 }
             }
             try{
-                database.execSQL("RELEASE cbl_" + Integer.toString(transactionLevel));
+                database.execSQL("RELEASE cbl_" + Integer.toString(tLevel));
             }
             catch (SQLException e) {
                 Log.e(Database.TAG, Thread.currentThread().getName() + " Error calling endTransaction()", e);
@@ -3447,10 +3462,11 @@ public final class Database {
 
     @InterfaceAudience.Private
     private void postChangeNotifications() {
+        int tLevel = transactionLevel.get();
         // This is a 'while' instead of an 'if' because when we finish posting notifications, there
         // might be new ones that have arrived as a result of notification handlers making document
         // changes of their own (the replicator manager will do this.) So we need to check again.
-        while (transactionLevel == 0 && isOpen() && !postingChangeNotifications
+        while (tLevel == 0 && isOpen() && !postingChangeNotifications
                 && changesToNotify.size() > 0)
         {
 
@@ -4892,5 +4908,6 @@ public final class Database {
         }
         return persistentCookieStore;
     }
+
 
 }

